### PR TITLE
fix backtick side effect in editor

### PR DESCRIPTION
**Pull request by tessus**
see https://github.com/laurent22/joplin/pull/1534#issuecomment-496310488 and following discussion.

This change is an ugly hack, which works. ace development will help react-ace to update ace, so I hope we can remove this soon. Until then...

I have tested it quite extensively, but I kindly ask that someone verifies the change.

### CODE REVIEW

**Comment by tessus** [comment]
also see https://github.com/ajaxorg/ace/issues/3954#issuecomment-496552563

**Comment by innocuo** [comment]
no disagreement that this is really ugly 😄 . Ace fix is 3 lines, Joplin fix is 111 lines.

Technically it seems correct as it's mostly the same code as the original csstyle behavior, but needs testing to make sure no other issues are introduced.

Comments:

- editor_ref() gets called about 6 times every time you type anything in the editor. Most of the code in editor_ref() should only execute once. So regardless if this new code is added, there should be a better way to make all the editor changes happen only once.

- how optimistic are we that react-ace will get updated? if it'll happen soon, why bother with hacks, and if it won't happen soon, we need a better long-term solution? (I'd be ok if this is 10 lines of code, but 100 feels a bit too much).

- Is it possible to create custom behaviors that extend CsStyle? If so, that could be in its own file, and the fix in this file might get reduced to `editor.getSession().getMode().$behavior = new JoplinBehavior()`

- Only comment on the code for now is to replace instances of `var` with `const/let` to keep the same code style as the rest of the file.

**Comment by tessus** [comment]
> Ace fix is 3 lines, Joplin fix is 111 lines.

I wish it wasn't, but it's the only way to fix this in userland.

> `editor_ref()` gets called about 6 times every time you type anything in the editor, there should be a better way to make all the editor changes happen only once.

Agreed, but as you mentioned, this has nothing to do with my ugly hack.

> how optimistic are we that react-ace will get updated? 

Good question. I have no idea. What would be a better long term solution though? I'm also not happy about the 100 lines. But either we apply this or revert the previous change, because this side effect might be too confusing for users.

> Is it possible to create custom behaviors that extend CsStyle?

Also a good question, which I do not know the answer to.

> Only comment on the code for now is to replace instances of var with const/let 

Since it's just a hack I didn't bother, especially if I use the wrong one, I will get errors/warnings and it will stop working.

**Comment by innocuo** [comment]
Ok, @tessus see my comments about the code, it looks like the hack can be a bit smaller, BUT, I don't know if that missing context var has some unintended consequences. I tested, and couldn't find any issues, but don't know enough about Ace to say it is 100% ok.

I think this one is for @laurent22 to decide if he wants this hack or rollback my previous PR for now.

**Comment by tessus** [comment]
@innocuo I don't see any comments about the code.

However, I can tell you that I started off with just the insertion function and I got errors. I had to add the getWrapped and initContext function. Then I also had to add the deletion part. It took me 2 hours to get it working and another 1.5 hours to test everything.

I can tell you that you can't remove any part of the code. If you do, it won't work. I only need someone to verify that the code works for them. I have tested for 1.5 hours, but I could have forgotten something.

I'm not looking for improvement of the code. It's a hack after all. I need people to verify it. that's all.

**Comment by innocuo** [comment]
> @innocuo I don't see any comments about the code.

I added that comment as code review.

> I'm not looking for improvement of the code. It's a hack after all. I need people to verify it. that's all.

I tested, editor seems to work ok, but it is setting the context var outside of the behavior, so it is never used by anything. So, technically, context var is being set where it shouldn't be set; I couldn't find any issues when testing, so it might be harmless.

**Comment by tessus** [comment]
> I added that comment as code review.

They might be still be pending, because I don't see them.

> setting the context var outside of the behavior, so it is never used by anything. So, technically, context var is being set where it shouldn't be set;

Try to comment the line

```
var context;
```

and see what happens. The code stops working. When you type a backtick character, you get a js error and the backtick character is not shown. (Not even a single backtick character.)

Anyway, I'm going to bed now. Will be back tomorrow.

**Comment by innocuo** [comment]
> They might be still pending, because I don't see them.

you are right, i just submitted it.
 
> > setting the context var outside of the behavior, so it is never used by anything. So, technically, context var is being set where it shouldn't be set;
> 
> Try to comment the line
> 
> ```
> var context;
> ```
> 
> and see what happens. The code stops working.

of course it does, because that's being used by initContext()
you can remove the two `initContext(editor);` calls in lines 926 and 983 and the code will still work (and when you remove that, you can remove context, contextCache and initContext). Because initContext just sets var context, and var context is being used nowhere in the new code, and the original behavior code doesn't have access to that var, so the whole initContext() part can be ignored. Or you can leave it.

File: ElectronClient/app/gui/NoteText.jsx [hunk 1]
```diff
@@ -880,9 +880,116 @@ class NoteTextComponent extends React.Component {
...
 				return output;
 			}
-			
-			//fixes #1426 but this is an Ace issue, so it can be removed if ace/brace is updated.
+
+			// fixes #1426 but this is an Ace issue, so it can be removed if ace/brace is updated.
+			// START of very ugly hack... (but it works)
+			var context;
+			var contextCache = {};
+			var initContext = function(editor) {
```
[the remark this answers is not in the document]
Comment (tessus) [reply ElectronClient/app/gui/NoteText.jsx]: Now I see what you are saying. Yes, makes sense. I've tried removing it locally, and I don't see any differences in behavior.

[the remark this answers is not in the document]
Comment (tessus) [reply ElectronClient/app/gui/NoteText.jsx]: Maybe @nightwing can answer this. He's been very helpful and understanding so far.

[the remark this answers is not in the document]
Comment (nightwing) [reply ElectronClient/app/gui/NoteText.jsx]: i think removing code related to context here will not cause any issues for markdown

[the remark this answers is not in the document]
Comment (tessus) [reply ElectronClient/app/gui/NoteText.jsx]: @nightwing thanks a lot for the info.

**Comment by innocuo** [comment]
ok, tested again on Mac, works ok for me.

**Comment by tessus** [comment]
Thanks, I've also tested all I could think of. I haven't seen any issues or side effects. All seems to be in order.

**Comment by laurent22** [comment]
That's quite heavy just for a relatively small feature, especially in this class which is already way too big. Would it make sense to wait if Ace update their lib? And perhaps we could work with react-ace so they get they package updated?

**Comment by tessus** [comment]
I agree, it is heavy. On the other side, it is the only way to mitigate the side effect that was introduced by 7dee93076a3d10f116961eaa703c725531776255.

Ace already fixed their code. It's rather the problem of getting react-ace and brace to follow suit.

**Comment by laurent22** [comment]
Alternatively we could also monkey patch whatever module is needed: https://github.com/kus/patch-module That way we move the code outside this main class.

File: ElectronClient/app/gui/NoteText.jsx [hunk 2]
```diff
@@ -880,9 +880,93 @@ class NoteTextComponent extends React.Component {
 				}
 				return output;
 			}
-			
-			//fixes #1426 but this is an Ace issue, so it can be removed if ace/brace is updated.
+
+			// fixes #1426 but this is an Ace issue, so it can be removed if ace/brace is updated.
+			// START of very ugly hack... (but it works)
+			var getWrapped = function(selection, selected, opening, closing) {
```
Comment (laurent22) [inline ElectronClient/app/gui/NoteText.jsx]: What is this getWrapped function for? It doesn't seem to be used so maybe we can remove it?

Comment (laurent22) [reply ElectronClient/app/gui/NoteText.jsx]: Edit: sorry it does get used.

**Comment by laurent22** [comment]
> I agree, it is heavy. On the other side, it is the only way to mitigate the side effect that was introduced by [7dee930](https://github.com/laurent22/joplin/commit/7dee93076a3d10f116961eaa703c725531776255).

I'd be keen to undo this change for now. It seems like so much effort for such a small feature (personally I didn't even know it existed before).

**Comment by tessus** [comment]
> I'd be keen to undo this change for now

I totally understand and I agree. (On the other side, it's just a stupid hack which would be removed as soon as upstream fixed their shit.) But I get it, I probably wouldn't want this mess in my code either.

But maybe we could make it easier for users. What about creating a shortcut that enclosed the selected text with backticks? e.g.: ``` Cmd+` ``` I could create a menu item in `Edit` after the item `Link`. What do you think?

In that case, we roll back 7dee930, but users will still be able to use the keyboard to enclose the selected text with a key combination. (I was wondering, why this was missing anyway.)
And as soon as upstream has included the Ace fix, we can either remove the shortcut again, leave it, or change it to create a code block instead.

**Comment by tessus** [comment]
@laurent, I wanted to follow up on this. I could work on this today.

- new PR
- revert 7dee930
- add shortcut ``` cmd+` ``` / ``` ctrl+` ```
- add menu item
- close this PR

Shall I do this or not?

**Comment by tessus** [comment]
@laurent22 (I used `@laurent` by mistake in my previous comment, so I do not know, if you got the notfication when I edited the comment and changed it to `@laurent22`)

**Comment by tessus** [comment]
@laurent22 see #1641

**Comment by tessus** [comment]
Closing this PR.

**Pull request closed without merging**